File: nvmolkit/fingerprints.py
# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""GPU-accelerated fingerprint generation."""

import torch

from nvmolkit._Fingerprints import MorganFingerprintGenerator as InternalFPGen
from nvmolkit.types import AsyncGpuResult, _resolve_cuda_stream


def unpack_fingerprint(fp: torch.Tensor) -> torch.Tensor:
    """Unpack a 32-bit integer-encoded fingerprint into a 2D boolean tensor of shape (len(fp), fingerprint_size).

    Args:
        fp: A tensor of shape `(n_fps, fp_size / 32)`, containing packed fingerprints with dtype uint32 or int32.

    Returns:
        A boolean tensor of shape `(n_fps, fp_size)`
    """
    if fp.ndim != 2:
        raise ValueError(f"Input tensor must be 2D, got shape={tuple(fp.shape)}")
    if fp.dtype not in (torch.int32, torch.uint32):
        raise ValueError("Input tensor must have dtype int32 or uint32")
    fp = fp.contiguous()
    if fp.dtype == torch.int32:
        fp = fp.view(torch.uint32)

    return (
        ((fp.to(torch.int64).unsqueeze(2) >> torch.arange(32, device=fp.device, dtype=torch.int64)) & 1)
        .bool()
        .flatten(1)
    )


def pack_fingerprint(fp: torch.Tensor) -> torch.Tensor:
    """Pack a 2D boolean tensor of shape `(n_fps, fingerprint_size)` into a 32-bit integer-encoded fingerprint.

    Args:
        fp: A boolean tensor of shape `(n_fps, fp_size)`

    Returns:
        A tensor of shape `(n_fps, fp_size / 32)` containing packed fingerprints (rounded up to the nearest multiple of 32)
    """
    if fp.ndim != 2:
        raise ValueError(f"Input tensor must be 2D, got shape={tuple(fp.shape)}")
    if fp.dtype != torch.bool:
        raise ValueError("Input tensor must have dtype bool")

    n_fps, fp_size = fp.shape
    padding = -fp_size % 32
    n_ints = (fp_size + padding) // 32

    # Pad to next multiple of 32 if needed
    if padding:
        padded = torch.zeros((n_fps, fp_size + padding), dtype=torch.bool, device=fp.device)
        padded[:, :fp_size] = fp
        fp = padded

    # Reshape to group bits into 32-bit chunks
    fp_reshaped = fp.reshape(n_fps, n_ints, 32)

    # Create powers of 2 for each bit position, using 0 to 31 instead of 31 to 0 to fix endianness
    powers = 1 << torch.arange(32, device=fp.device, dtype=torch.int64)

    # Multiply and sum to create packed integers
    return (fp_reshaped * powers.unsqueeze(0)).sum(dim=2).to(torch.uint32)


class MorganFingerprintGenerator:
    """Morgan fingerprint generator."""

    def __init__(self, radius: int, fpSize: int):
        """Initialize the Morgan fingerprint generator.

        Args:
            radius: The radius of the Morgan fingerprint.
            fpSize: The size of the fingerprint. Must be one of {128, 256, 512, 1024, 2048}.
        """
        self._internal = InternalFPGen(radius, fpSize)

    def GetFingerprints(
        self, mols: list, num_threads: int = 0, stream: torch.cuda.Stream | None = None
    ) -> AsyncGpuResult:
        """Compute Morgan fingerprints for a list of molecules.

        Preprocessing of fingerprinting features is done on the CPU, and is parallelized with the `num_threads` argument.
        The resulting tensor has dtype torch.uint32 and contains a packed fingerprint for each molecule, one row per
        molecule.

        Packed fingerprints can be passed directly to nvMolKit similarity calculations, or unpacked
        via `unpack_fingerprint`.

        Args:
            mols: List of RDKit molecules to generate fingerprints for
            num_threads: Number of CPU threads to use for fingerprint generation. If 0, uses all available threads.
            stream: CUDA stream to use. If None, uses the current stream.

        Returns:
            AsyncGpuResult wrapping a torch.Tensor of shape (len(mols), fpSize / 32) containing the fingerprints.
            Each row is a fingerprint for the corresponding molecule.
        """
        active_stream = _resolve_cuda_stream(stream)
        return AsyncGpuResult(self._internal.GetFingerprintsDevice(mols, num_threads, active_stream.cuda_stream))

File: nvmolkit/fingerprints.cpp
// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <DataStructs/ExplicitBitVect.h>
#include <GraphMol/ROMol.h>

#include <boost/python.hpp>

#include "nvmolkit/array_helpers.h"
#include "src/morgan_fingerprint.h"
#include "src/utils/device.h"
#include "src/utils/nvtx.h"

namespace {

using namespace boost::python;

struct PythonMoleculeSequence {
  object                           owner;
  std::vector<const RDKit::ROMol*> molecules;
};

PythonMoleculeSequence convertMolecules(const object& mols) {
  if (!PySequence_Check(mols.ptr())) {
    PyErr_SetString(PyExc_TypeError, "mols must be a sequence of RDKit molecules");
    throw_error_already_set();
  }
  PyObject* sequence = PySequence_Fast(mols.ptr(), "mols must be a sequence of RDKit molecules");
  if (sequence == nullptr) {
    throw_error_already_set();
  }

  PythonMoleculeSequence result{object(handle<>(sequence)), {}};
  const Py_ssize_t       numMols = PySequence_Fast_GET_SIZE(sequence);
  PyObject* const*       items   = PySequence_Fast_ITEMS(sequence);
  result.molecules.reserve(static_cast<std::size_t>(numMols));
  for (Py_ssize_t i = 0; i < numMols; ++i) {
    const RDKit::ROMol* mol = extract<const RDKit::ROMol*>(items[i]);
    if (mol == nullptr) {
      throw std::invalid_argument("Invalid molecule at index " + std::to_string(i));
    }
    result.molecules.push_back(mol);
  }
  return result;
}

template <int nBits>
nvMolKit::PyArray* makePyArrayFromFlatBitVects(nvMolKit::AsyncDeviceVector<nvMolKit::FlatBitVect<nBits>>& deviceVect) {
  using dtype                = typename nvMolKit::FlatBitVect<nBits>::StorageType;
  const std::string dTypeStr = nvMolKit::getNumpyType<dtype>();

  // Make a 2D array, rows are the number of fingerprints, columns are the number of block_types
  const int nRows = deviceVect.size();
  const int nCols = nBits / (8 * sizeof(dtype));

  return nvMolKit::makePyArray(deviceVect, dTypeStr, boost::python::make_tuple(nRows, nCols));
}

template <int nBits>
nvMolKit::PyArray* getFingerprintsDevice(nvMolKit::MorganFingerprintGenerator&      generator,
                                         const std::vector<const RDKit::ROMol*>&    mols,
                                         cudaStream_t                               stream,
                                         const nvMolKit::FingerprintComputeOptions& computeOptions) {
  auto fingerprints = [&]() {
    nvMolKit::ScopedNvtxRange range("MorganFPBindingNativeCompute", nvMolKit::NvtxColor::kOrange);
    return generator.GetFingerprintsGpuBuffer<nBits>(mols, stream, computeOptions);
  }();

  nvMolKit::ScopedNvtxRange range("MorganFPBindingOutputWrapping", nvMolKit::NvtxColor::kCyan);
  return makePyArrayFromFlatBitVects<nBits>(fingerprints);
}

}  // namespace

BOOST_PYTHON_MODULE(_Fingerprints) {
  class_<nvMolKit::MorganFingerprintGenerator, boost::noncopyable>(
    "MorganFingerprintGenerator",
    init<const std::uint32_t, const std::uint32_t>((boost::python::arg("radius"), boost::python::arg("fpSize"))))
    .def(
      "GetFingerprint",
      +[](nvMolKit::MorganFingerprintGenerator& selfref, const RDKit::ROMol& mol) {
        return selfref.GetFingerprint(mol).release();
      },
      return_value_policy<manage_new_object>())
    .def(
      "GetFingerprintsDevice",
      +[](nvMolKit::MorganFingerprintGenerator& selfref,
          const boost::python::object&          mols,
          int                                   numThreads,
          std::uintptr_t                        streamPtr) {
        auto convertedMols = [&]() {
          nvMolKit::ScopedNvtxRange range("MorganFPBindingInputConversion", nvMolKit::NvtxColor::kYellow);
          return convertMolecules(mols);
        }();

        nvMolKit::FingerprintComputeOptions computeOptions;
        computeOptions.backend       = nvMolKit::FingerprintComputeBackend::GPU;
        computeOptions.numCpuThreads = numThreads;
        auto streamOpt               = [&]() {
          nvMolKit::ScopedNvtxRange range("MorganFPBindingStreamAcquisition", nvMolKit::NvtxColor::kBlue);
          return nvMolKit::acquireExternalStream(streamPtr);
        }();
        if (!streamOpt) {
          throw std::invalid_argument("Invalid CUDA stream");
        }
        auto        stream  = *streamOpt;
        const auto& options = selfref.GetOptions();
        switch (options.fpSize) {
          case 128: {
            return getFingerprintsDevice<128>(selfref, convertedMols.molecules, stream, computeOptions);
          }
          case 256: {
            return getFingerprintsDevice<256>(selfref, convertedMols.molecules, stream, computeOptions);
          }
          case 512: {
            return getFingerprintsDevice<512>(selfref, convertedMols.molecules, stream, computeOptions);
          }
          case 1024: {
            return getFingerprintsDevice<1024>(selfref, convertedMols.molecules, stream, computeOptions);
          }
          case 2048: {
            return getFingerprintsDevice<2048>(selfref, convertedMols.molecules, stream, computeOptions);
          }
          default:
            throw std::invalid_argument("Invalid fpSize: " + std::to_string(options.fpSize) +
                                        ". Supported values are 128, 256, 512, 1024, 2048");
        }
      },
      return_value_policy<manage_new_object>(),
      (boost::python::arg("mols"), boost::python::arg("num_threads") = 0, boost::python::arg("stream") = 0));
}
